File: study/lua/src/app/scene/view/silkbag/PopupSilkbagDetailEx.lua
-- 锦囊详情弹框
-- Author: Liangxu
-- 
local PopupBase = require("app.ui.PopupBase")
local PopupSilkbagDetailEx = class("PopupSilkbagDetailEx", PopupBase)
local SilkbagDataHelper = require("app.utils.data.SilkbagDataHelper")
local SilkbagDetailExCell = require("app.scene.view.silkbag.SilkbagDetailExCell")
local TypeConvertHelper = require("app.utils.TypeConvertHelper")
local HeroDataHelper = require("app.utils.data.HeroDataHelper")
local SilkbagConst = require("app.const.SilkbagConst")

local Color2Res = {
	[4] = "img_bg_silkbag_01",
	[5] = "img_bg_silkbag_02",
	[6] = "img_bg_silkbag_03",
	[7] = "img_bg_silkbag_04",
}

function PopupSilkbagDetailEx:ctor(type, value)
	self._type = type
	self._value = value
	self._panelRoot = nil

	local resource = {
		file = Path.getCSB("PopupSilkbagDetailEx", "silkbag"),
		binding = {
			_btnWayGet = {
				events = {{event = "touch", method = "_onButtonWayGetClicked"}}
			},
			_buttonClose = {
				events = {{event = "touch", method = "_onButtonClose"}}
			},
		}
	}
	PopupSilkbagDetailEx.super.ctor(self, resource)
end

function PopupSilkbagDetailEx:onCreate()
	self._silkId = 0
	self._heroBaseIds = {}

	self._fileNodeSilkbag:setVisible(true)
	self._scrollPage:setVisible(false)
	self._btnWayGet:setString(Lang.get("way_type_goto_get"))
	self._fileNodeTitle:setTitle(Lang.get("silkbag_suit_hero_title"))

	self._listView:setTemplate(SilkbagDetailExCell)
	self._listView:setCallback(handler(self, self._onItemUpdate), handler(self, self._onItemSelected))
	self._listView:setCustomCallback(handler(self, self._onItemTouch))
	self:_initDesc()
end

function PopupSilkbagDetailEx:onEnter()
	if G_ConfigManager:isDalanVersion() then
		self._nodeEmpty:getSubNodeByName("ImageWaterFlow"):setVisible(false)
	end
	self:_updateUI(self._value)
end

function PopupSilkbagDetailEx:onExit()
	if self._callBack then
		self._callBack()
	end
end

function PopupSilkbagDetailEx:setCloseCallBack(callback)
	self._callBack = callback
	G_SignalManager:dispatch(SignalConst.EVENT_SEASONSPORT_OPENSILKDETAIL)
end


function PopupSilkbagDetailEx:_updateDesc(strDesc)
	self._textDesTrue:setString(strDesc)
	
	local labelSize = self._textDesTrue:getContentSize()
	self._textDescSc:setInnerContainerSize(labelSize)
	local orgHeight = self._textDescSc:getContentSize().height
	local height = math.max(orgHeight, labelSize.height)
	if math.abs(labelSize.width-self._textDes:getContentSize().width)>10 then 	-- 不满1行，位置调整
		local x = (self._textDes:getContentSize().width-labelSize.width)*0.5
		self._textDesTrue:setPosition(cc.p(x, height))
	else
		self._textDesTrue:setPosition(cc.p(0, height))
	end
	local enable = labelSize.height>orgHeight
	self._textDescSc:setTouchEnabled(enable)
	self._textDescSc:setSwallowTouches(enable)
end

function PopupSilkbagDetailEx:_initDesc()
	local size = self._textDes:getContentSize()
	local sc = ccui.ScrollView:create()
	sc:setBounceEnabled(true)
	sc:setDirection(ccui.ScrollViewDir.vertical)
	sc:setTouchEnabled(true)
	sc:setSwallowTouches(true)
	sc:setScrollBarEnabled(false)
	sc:setContentSize(size)

	local label = cc.Label:createWithTTF("", Path.getCommonFont(), self._textDes:getFontSize())
	label:setColor(self._textDes:getColor())
	label:setMaxLineWidth(size.width)
	-- label:setAnchorPoint(self._textDes:getAnchorPoint())
	label:setAnchorPoint(cc.p(0, 1))
	self._textDesTrue = label

	self._textDes:getParent():addChild(sc)
	sc:setAnchorPoint(self._textDes:getAnchorPoint())
	sc:setPosition(cc.p(self._textDes:getPosition()))
	self._textDes:setVisible(false)
	sc:addChild(label)
	self._textDescSc = sc
end

function PopupSilkbagDetailEx:_updateUI(baseId)
	self._value = baseId
	local silkbagBaseId = self._value
	self._silkId = SilkbagDataHelper.getSilkbagConfig(silkbagBaseId).mapping

	local info = SilkbagDataHelper.getSilkbagConfig(silkbagBaseId)
	local strPower = "+"..info.fake
	local colorBgRes = Color2Res[info.color]

	self._textPower:setString(strPower)
	if colorBgRes then
		self._imageColorBg:loadTexture(Path.getBackground(colorBgRes, ".png"))	
	end
	self._fileNodeSilkbag:updateUI(silkbagBaseId)
	-- self._textDes:setString(info.description)
	self:_updateDesc(info.description)

	self:_updateListView(silkbagBaseId)
end

function PopupSilkbagDetailEx:_updateListView(silkbagBaseId)
	local heroBaseIds, suitType = G_UserData:getSilkbag():getHeroIdsWithSilkbagId(silkbagBaseId)
	if suitType == SilkbagConst.SUIT_TYPE_NONE then
		self._nodeEmpty:setVisible(false)
		self._nodeList:setVisible(true)
		self._heroBaseIds = self:_filterHeroIds(heroBaseIds)
		self._listView:clearAll()
	    self._listView:resize(#self._heroBaseIds)
	else
		self._nodeEmpty:setVisible(true)
		self._nodeList:setVisible(false)
		if suitType == SilkbagConst.SUIT_TYPE_ALL then
			self._textEmptyTip:setString(Lang.get("silkbag_suit_tip_all"))
		elseif suitType == SilkbagConst.SUIT_TYPE_MALE then
			self._textEmptyTip:setString(Lang.get("silkbag_suit_tip_male"))
		elseif suitType == SilkbagConst.SUIT_TYPE_FEMALE then
			self._textEmptyTip:setString(Lang.get("silkbag_suit_tip_female"))
		end
	end
end

function PopupSilkbagDetailEx:_filterHeroIds(heroBaseIds)
	local isLeaderExist = false
	local temp = {}
	local result = {}

	local function sortFunc(a, b)
		if a.type ~= b.type then
			return a.type < b.type
		elseif a.color ~= b.color then
			return a.color > b.color
		else
			return a.id < b.id
		end
	end

	local gender = G_UserData:getBase():isMale() and 1 or 2
	for i, heroBaseId in ipairs(heroBaseIds) do
		local info = HeroDataHelper.getHeroConfig(heroBaseId)
		if info.type == 1 then --主角
			if info.gender == gender and not isLeaderExist then
				table.insert(temp, info)
				isLeaderExist = true
			end
		else
			table.insert(temp, info)
		end
	end

	table.sort(temp, sortFunc)
	for i, data in ipairs(temp) do
		table.insert(result, data.id)
	end

	return result
end

function PopupSilkbagDetailEx:_onItemUpdate(item, index)
	local index = index + 1
	local heroBaseId = self._heroBaseIds[index]
	if heroBaseId then
		item:update(self._silkId, heroBaseId)
	end
end

function PopupSilkbagDetailEx:_onItemSelected(item, index)
	
end

function PopupSilkbagDetailEx:_onItemTouch(index, t)
    
end

function PopupSilkbagDetailEx:_onButtonWayGetClicked()
	local popup = require("app.ui.PopupItemGuider").new(Lang.get("way_type_get"))
	popup:updateUI(TypeConvertHelper.TYPE_SILKBAG, self._value)
	popup:openWithAction()
end

function PopupSilkbagDetailEx:_onButtonClose()
	self:close()
end

function PopupSilkbagDetailEx:setPageData(dataList, params)
	self._dataList = dataList
	self._scrollPage:setCallBack(handler(self, self._updateItemAvatar))
	self._scrollPage:setVisible(true)
	self._fileNodeSilkbag:setVisible(false)
	local selectPos = 0
	for i, data in ipairs(self._dataList) do
		if  data.cfg.id == self._value then
			selectPos = i
		end
	end
	self._scrollPage:setUserData(dataList, selectPos)
end

function PopupSilkbagDetailEx:_updateItemAvatar(sender, widget, index, selectPos)
	local data = self._dataList[index]
	if data == nil then
		return
	end
	local baseId = data.cfg.id

	local count = widget:getChildrenCount()
	if count == 0 then
		local CSHelper = require("yoka.utils.CSHelper")
		local avatar = CSHelper.loadResourceNode(Path.getCSB("CommonSilkbagAvatar", "common"))
		avatar:updateUI(baseId)
		avatar:setPosition(cc.p(self._scrollPage:getPageSize().width / 2, self._scrollPage:getPageSize().height / 2))
		widget:addChild(avatar)
	end

	if selectPos == index then
		self:_updateUI(baseId)
	end

end

-- @Role 	如果在无差别竞技中调用需要特殊处理
function PopupSilkbagDetailEx:updateInSeasonSilkView()
	self._btnWayGet:setVisible(false)
	self._panelRoot:setScale(0.9)
end

return PopupSilkbagDetailEx

File: study/ts/script/scene/view/silkbag/PopupSilkbagDetailEx.ts
const { ccclass, property } = cc._decorator;

import CommonButtonLevel5Normal from '../../../ui/component/CommonButtonLevel5Normal'

import CommonPageItemSilkbag from '../../../ui/component/CommonPageItemSilkbag'

import CommonSilkbagAvatar from '../../../ui/component/CommonSilkbagAvatar'

import CommonEmptyListNode from '../../../ui/component/CommonEmptyListNode'

import CommonDetailTitleWithBg from '../../../ui/component/CommonDetailTitleWithBg'
import CommonListView from '../../../ui/component/CommonListView';
import { Lang } from '../../../lang/Lang';
import SilkbagDetailExCell from './SilkbagDetailExCell';
import { handler } from '../../../utils/handler';
import { G_SignalManager, G_UserData } from '../../../init';
import { SignalConst } from '../../../const/SignalConst';
import { SilkbagDataHelper } from '../../../utils/data/SilkbagDataHelper';
import { Path } from '../../../utils/Path';
import { SilkbagConst } from '../../../const/SilkbagConst';
import { HeroDataHelper } from '../../../utils/data/HeroDataHelper';
import { TypeConvertHelper } from '../../../utils/TypeConvertHelper';
import PopupBase from '../../../ui/PopupBase';
import UIHelper from '../../../utils/UIHelper';
import { UIPopupHelper } from '../../../utils/UIPopupHelper';

var Color2Res = {
    4: 'img_bg_silkbag_01',
    5: 'img_bg_silkbag_02',
    6: 'img_bg_silkbag_03',
    7: 'img_bg_silkbag_04'
};

@ccclass
export default class PopupSilkbagDetailEx extends PopupBase {
    public static path = 'silkbag/PopupSilkbagDetailEx';
    @property({
        type: cc.Node,
        visible: true
    })
    _panelRoot: cc.Node = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textTitle: cc.Label = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _nodeList: cc.Node = null;

    @property({
        type: CommonDetailTitleWithBg,
        visible: true
    })
    _fileNodeTitle: CommonDetailTitleWithBg = null;

    @property({
        type: CommonListView,
        visible: true
    })
    _listView: CommonListView = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _nodeEmpty: cc.Node = null;

    @property({
        type: CommonEmptyListNode,
        visible: true
    })
    _imageWaterFlow: CommonEmptyListNode = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textEmptyTip: cc.Label = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageColorBg: cc.Sprite = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textPower: cc.Label = null;

    @property({
        type: CommonSilkbagAvatar,
        visible: true
    })
    _fileNodeSilkbag: CommonSilkbagAvatar = null;

    @property({
        type: CommonPageItemSilkbag,
        visible: true
    })
    _scrollPage: CommonPageItemSilkbag = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageBtnBg: cc.Sprite = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textDes: cc.Label = null;

    @property({
        type: CommonButtonLevel5Normal,
        visible: true
    })
    _btnWayGet: CommonButtonLevel5Normal = null;

    @property({
        type: cc.Button,
        visible: true
    })
    _buttonClose: cc.Button = null;
    _type: any;
    _value: any;
    _silkId: number;
    _heroBaseIds;
    _callBack: any;

    ctor(type, value) {
        this._type = type;
        this._value = value;
    }
    onCreate() {
        this._silkId = 0;
        this._heroBaseIds = {};
        this._fileNodeSilkbag.node.active = (true);
        this._scrollPage.node.active = (false);
        this._btnWayGet.setString(Lang.get('way_type_goto_get'));
        this._fileNodeTitle.setTitle(Lang.get('silkbag_suit_hero_title'));
        this._btnWayGet.addClickEventListenerEx(handler(this, this._onButtonWayGetClicked));

    }
    onEnter() {
        this._listView.init(null, handler(this, this._onItemUpdate));
        // if (G_ConfigManager.isDalanVersion()) {
        //     this._nodeEmpty.getChildByName('ImageWaterFlow')..active = (false);
        // }
        this._updateUI(this._value);
    }
    onExit() {
        if (this._callBack) {
            this._callBack();
        }
    }
    setCloseCallBack(callback) {
        this._callBack = callback;
        G_SignalManager.dispatch(SignalConst.EVENT_SEASONSPORT_OPENSILKDETAIL);
    }
    _updateUI(baseId) {
        this._value = baseId;
        var silkbagBaseId = this._value;
        this._silkId = SilkbagDataHelper.getSilkbagConfig(silkbagBaseId).mapping;
        var info = SilkbagDataHelper.getSilkbagConfig(silkbagBaseId);
        var strPower = '+' + info.fake;
        var colorBgRes = Color2Res[info.color];
        this._textPower.string = (strPower);
        if (colorBgRes) {
            UIHelper.loadTexture(this._imageColorBg, Path.getBackground(colorBgRes));
        }
        this._fileNodeSilkbag.updateUI(silkbagBaseId);
        this._textDes.string = (info.description);
        this._updateListView(silkbagBaseId);
    }
    _updateListView(silkbagBaseId) {
        var [heroBaseIds,suitType] = G_UserData.getSilkbag().getHeroIdsWithSilkbagId(silkbagBaseId);
        if (suitType == SilkbagConst.SUIT_TYPE_NONE) {
            this._nodeEmpty.active = (false);
            this._nodeList.active = (true);
            this._heroBaseIds = this._filterHeroIds(heroBaseIds);
            this._listView.setData(this._heroBaseIds.length);
        } else {
            this._nodeEmpty.active = (true);
            this._nodeList.active = (false);
            if (suitType == SilkbagConst.SUIT_TYPE_ALL) {
                this._textEmptyTip.string = (Lang.get('silkbag_suit_tip_all'));
            } else if (suitType == SilkbagConst.SUIT_TYPE_MALE) {
                this._textEmptyTip.string = (Lang.get('silkbag_suit_tip_male'));
            } else if (suitType == SilkbagConst.SUIT_TYPE_FEMALE) {
                this._textEmptyTip.string = (Lang.get('silkbag_suit_tip_female'));
            }
        }
    }
    _filterHeroIds(heroBaseIds) {
        var isLeaderExist = false;
        var temp = [];
        var result = [];
        function sortFunc(a, b) {
            if (a.type != b.type) {
                return a.type - b.type;
            } else if (a.color != b.color) {
                return b.color - a.color;
            } else {
                return a.id - b.id;
            }
        }
        var gender = G_UserData.getBase().isMale() ? 1 : 2;
        for (var i in heroBaseIds) {
            var heroBaseId = heroBaseIds[i];
            var info = HeroDataHelper.getHeroConfig(heroBaseId);
            if (info.type == 1) {
                if (info.gender == gender && !isLeaderExist) {
                    temp.push(info);
                    isLeaderExist = true;
                }
            } else {
                temp.push(info);
            }
        }
        temp.sort(sortFunc);
        for (i in temp) {
            var data = temp[i];
            result.push(data.id);
        }
        return result;
    }
    _onItemUpdate(item, index, type) {
        var index = index;
        var heroBaseId = this._heroBaseIds[index];
        if (heroBaseId) {
            item.updateItem(index, [this._silkId, heroBaseId], type);
        }else {
            item.updateItem(index, null, type);
        }
    }
    _onItemSelected(item, index) {
    }
    _onItemTouch(index, t) {
    }
    _onButtonWayGetClicked() {
        UIPopupHelper.popupItemGuiderByType(TypeConvertHelper.TYPE_SILKBAG, this._value, Lang.get("way_type_get"))
    }
    onButtonClose() {
        this.close();
    }
    setPageData(dataList, params) {
        // this._dataList = dataList;
        // this._scrollPage.setCallBack(handler(this, this._updateItemAvatar));
        // this._scrollPage.node.active = (true);
        // this._fileNodeSilkbag.node.active = (false);
        // var selectPos = 0;
        // for (i in this._dataList) {
        //     var data = this._dataList[i];
        //     if (data.cfg.id == this._value) {
        //         selectPos = i;
        //     }
        // }
        // this._scrollPage.setUserData(dataList, selectPos);
    }
    _updateItemAvatar(sender, widget, index, selectPos) {
        // var data = this._dataList[index];
        // if (data == null) {
        //     return;
        // }
        // var baseId = data.cfg.id;
        // var count = widget.getChildrenCount();
        // if (count == 0) {
        //     var CSHelper = require('CSHelper');
        //     var avatar = CSHelper.loadResourceNode(Path.getCSB('CommonSilkbagAvatar', 'common'));
        //     avatar.updateUI(baseId);
        //     avatar.setPosition(cc.v2(this._scrollPage.getPageSize().width / 2, this._scrollPage.getPageSize().height / 2));
        //     widget.addChild(avatar);
        // }
        // if (selectPos == index) {
        //     this._updateUI(baseId);
        // }
    }
    updateInSeasonSilkView() {
        this._btnWayGet.node.active = (false);
        this._panelRoot.setScale(0.9);
    }

}